src/main/java/gregtech/api/util/GTTransferUtils.java

```diff
@@ -0,0 +1,250 @@
+package gregtech.api.util;
+
+import gregtech.api.capability.IMultipleTankHandler;
+import gregtech.api.recipes.FluidKey;
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidTankProperties;
+import net.minecraftforge.items.IItemHandler;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+public class GTTransferUtils {
+
+    public static int transferFluids(@Nonnull IFluidHandler sourceHandler, @Nonnull IFluidHandler destHandler, int transferLimit) {
+        return transferFluids(sourceHandler, destHandler, transferLimit, fluidStack -> true);
+    }
+
+    public static int transferFluids(@Nonnull IFluidHandler sourceHandler, @Nonnull IFluidHandler destHandler, int transferLimit, @Nonnull Predicate<FluidStack> fluidFilter) {
+        int fluidLeftToTransfer = transferLimit;
+
+        for (IFluidTankProperties tankProperties : sourceHandler.getTankProperties()) {
+            FluidStack currentFluid = tankProperties.getContents();
+            if (currentFluid == null || currentFluid.amount == 0 || !fluidFilter.test(currentFluid)) {
+                continue;
+            }
+
+            currentFluid.amount = fluidLeftToTransfer;
+            FluidStack fluidStack = sourceHandler.drain(currentFluid, false);
+            if (fluidStack == null || fluidStack.amount == 0) {
+                continue;
+            }
+
+            int canInsertAmount = destHandler.fill(fluidStack, false);
+            if (canInsertAmount > 0) {
+                fluidStack.amount = canInsertAmount;
+                fluidStack = sourceHandler.drain(fluidStack, true);
+                if (fluidStack != null && fluidStack.amount > 0) {
+                    destHandler.fill(fluidStack, true);
+
+                    fluidLeftToTransfer -= fluidStack.amount;
+                    if (fluidLeftToTransfer == 0) {
+                        break;
+                    }
+                }
+            }
+        }
+        return transferLimit - fluidLeftToTransfer;
```

Comment:
It is redundant to loop through `getTankProperties` since `fill` and `drain` explicitly describes those operations to affect all *fluid tanks* in the `IFluidHandler`. By definition, an `IFluidHandler` impl's `fill` and `drain` operations would loop through tanks itself anyway.

Reply:
I do not see a good way to rewrite this method. `IFluidHandler` only contains `drain(FluidStack fluid, boolean doDrain)` an `drain(int maxAmount, boolean doDrain)`

Reply:
The method without the predicate should be able to be optimized, I see why you have to loop through `getTankProperties` for the predicate version of the method now.